src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/CodeStyleOptions2.cs

```diff
@@ -377,6 +377,56 @@ private static string GetAccessibilityModifiersRequiredEditorConfigString(CodeSt
                 KeyValuePairUtil.Create("all", UnusedParametersPreference.AllMethods),
             });
 
+        #region dotnet_style_prefer_foreach_explicit_cast_in_Source
+
+        private static readonly CodeStyleOption2<ForEachExplicitCastInSourcePreference> s_forEachExplicitCastInSourceNonLegacyPreference =
+            new(ForEachExplicitCastInSourcePreference.NonLegacy, NotificationOption2.Suggestion);
+
+        private static readonly BidirectionalMap<string, ForEachExplicitCastInSourcePreference> s_forEachExplicitCastInSourcePreferencePreferenceMap =
+            new(new[]
+            {
+                KeyValuePairUtil.Create("always", ForEachExplicitCastInSourcePreference.Always),
+                KeyValuePairUtil.Create("never", ForEachExplicitCastInSourcePreference.Never),
+                KeyValuePairUtil.Create("non_legacy", ForEachExplicitCastInSourcePreference.NonLegacy),
+            });
+
+        internal static readonly PerLanguageOption2<CodeStyleOption2<ForEachExplicitCastInSourcePreference>> ForEachExplicitCastInSource =
+            CreateOption(
+                CodeStyleOptionGroups.ExpressionLevelPreferences,
+                nameof(ForEachExplicitCastInSource),
+                s_forEachExplicitCastInSourceNonLegacyPreference,
+                new EditorConfigStorageLocation<CodeStyleOption2<ForEachExplicitCastInSourcePreference>>(
+                    "dotnet_style_prefer_foreach_explicit_cast_in_source",
+                    s => ParseForEachExplicitCastInSourcePreference(s, s_forEachExplicitCastInSourceNonLegacyPreference),
+                    v => GetForEachExplicitCastInSourceEditorConfigString(v, s_forEachExplicitCastInSourceNonLegacyPreference)),
+                new RoamingProfileStorageLocation($"TextEditor.%LANGUAGE%.Specific.ForEachExplicitCastInSourcePreference"));
+
+        private static CodeStyleOption2<ForEachExplicitCastInSourcePreference> ParseForEachExplicitCastInSourcePreference(
+            string optionString, CodeStyleOption2<ForEachExplicitCastInSourcePreference> defaultValue)
+        {
+            if (TryGetCodeStyleValueAndOptionalNotification(optionString,
+                    defaultValue.Notification, out var value, out var notification))
+            {
+                Debug.Assert(s_forEachExplicitCastInSourcePreferencePreferenceMap.ContainsKey(value));
+                return new CodeStyleOption2<ForEachExplicitCastInSourcePreference>(
+                    s_forEachExplicitCastInSourcePreferencePreferenceMap.GetValueOrDefault(value), notification);
+            }
+
+            return defaultValue;
+        }
+
+        private static string GetForEachExplicitCastInSourceEditorConfigString(
+            CodeStyleOption2<ForEachExplicitCastInSourcePreference> option,
+            CodeStyleOption2<ForEachExplicitCastInSourcePreference> defaultValue)
+        {
+            Debug.Assert(s_forEachExplicitCastInSourcePreferencePreferenceMap.ContainsValue(option.Value));
```

Comment:
What if the user provided an invalid value in editorconfig?

Reply:
we debug assert :)  but it doesn't hit anyone at runtime.  this helps us capture issues in internal debug builds that are more likely to be a code issue.  but in release this doesn't happen since clearly it could happen with a user.